### ISSUE TITLE
PGroonga 1.2.0 crash followed by Postgres crash/recovery

### ISSUE THREAD

**Issue by peter-schmitz** [pgroonga/pgroonga#36]
Since the last update on 2017-05-01, pgroonga / groonga crashes regulary.
Basic Information:
- Postgres 9.5.4 on Ubuntu 14.04 x64
- PGroonga 1.2.0-2~xenial1
- libgroonga 7.0.2-2~xenial1
- Accessing Postgres via JDBC.
- happens up to 6 times per day, at least once on a machine with ECC ram.
- Only 2 PGroonga indexes in use:
create index groonga1 on searchentry using pgroonga( fulltext1 ) WITH
(tokenizer='TokenTrigram', normalizer='');
( ... groonga2 ... fulltext2, same definition otherwise )
- Did not occur with PGroonga 1.1.9 and libgroonga 7.0.1

pgroonga.log:

```text
2017-05-04 13:51:30.916892|i|8313: hits=40
2017-05-04 13:51:30.920866|i|8378: [object][search][index][key][exact] <Lexicon11521365_0.index>
2017-05-04 13:51:30.920884|i|8378: grn_ii_sel > (181956)
2017-05-04 13:51:30.920925|i|8378: n=4 (181956)
2017-05-04 13:51:30.923152|d|8378: [ii][cursor][min] skip: 0x5630de53dde0: min(1222078->1303320): chunk(33->36): chunk-used(true->true)
2017-05-04 13:51:30.923530|d|8378: [ii][cursor][min] skip: 0x5630de53dde0: min(1479703->1501695): chunk(43->44): chunk-used(true->true)
2017-05-04 13:51:30.925069|i|8378: exact: 40
2017-05-04 13:51:30.925079|i|8378: hits=40
2017-05-04 13:51:30.925221|i|8378: [object][search][index][key][exact] <Lexicon11521366_0.index>
2017-05-04 13:51:30.925227|i|8378: grn_ii_sel > (181956)
2017-05-04 13:51:30.925274|i|8378: n=4 (181956)
2017-05-04 13:51:30.931706|i|8378: exact: 53
2017-05-04 13:51:30.931726|i|8378: hits=53
2017-05-04 13:51:41.947700|n|11095: grn_init: <7.0.2>
2017-05-04 13:51:41.947824|n|11095: pgroonga: initialize: <1.2.0>
...
2017-05-04 13:52:32.148899|i|11194: <0x7f115e4822a0:0> expired
i=0x5630df1984d0 max=625 (3/3)
2017-05-04 13:52:32.190654|d|11194: flushing a[0]=460828
seg=7(0x7f114e1d4000) free=24
2017-05-04 13:52:32.203047|d|11194: flushed  a[0]=460828
seg=7(0x7f114b6d4000) free=24->247712 nterms=901 v=1
2017-05-04 13:52:32.203176|i|11194: <0x7f115e4822a0:0> expired
i=0x5630df1984d0 max=625 (3/3)
2017-05-04 13:52:35.206077|C|8378: -- CRASHED!!! --
2017-05-04 13:52:35.213660|C|8378:
/usr/lib/x86_64-linux-gnu/libgroonga.so.0(+0xa0e9e) [0x7f115d85be9e]
2017-05-04 13:52:35.213688|C|8378:
/lib/x86_64-linux-gnu/libc.so.6(+0x36cb0) [0x7f138aa0bcb0]
2017-05-04 13:52:35.213693|C|8378:
/usr/lib/postgresql/9.5/lib/pgroonga.so(+0x7e01) [0x7f115e05ee01]
2017-05-04 13:52:35.213696|C|8378:
/usr/lib/postgresql/9.5/lib/pgroonga.so(+0x7ec7) [0x7f115e05eec7]
2017-05-04 13:52:35.213699|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(+0x2d98a1) [0x5630db6918a1]
2017-05-04 13:52:35.213703|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(proc_exit+0x10) [0x5630db6918e0]
2017-05-04 13:52:35.213706|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(PostgresMain+0x41f) [0x5630db6b01ef]
2017-05-04 13:52:35.213710|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(+0x9d678) [0x5630db455678]
2017-05-04 13:52:35.213713|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(PostmasterMain+0x104b) [0x5630db65742b]
2017-05-04 13:52:35.213716|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(main+0x842) [0x5630db456372]
2017-05-04 13:52:35.213719|C|8378:
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7f138a9f6f45]
2017-05-04 13:52:35.213722|C|8378: postgres: postgres hostname
127.0.0.1(57248) idle(+0x9e3dd) [0x5630db4563dd]
2017-05-04 13:52:35.213726|C|8378: ----------------
2017-05-04 13:52:35.218067|n|8313: grn_fin (0)
2017-05-04 13:52:35.223029|n|8283: grn_fin (0)
2017-05-04 13:52:57.134642|n|25104: grn_init: <7.0.2>
2017-05-04 13:52:57.134779|n|25104: pgroonga: initialize: <1.2.0>
```

From postgresql.log:

```text
2017-05-04 13:52:35 CEST [1920-39] LOG:  Serverprozess (PID 8378) wurde
von Signal 6 beendet: Aborted
2017-05-04 13:52:35 CEST [1920-40] LOG:  aktive Serverprozesse werden
abgebrochen
2017-05-04 13:52:35 CEST [11095-3] postgres@hostname WARNUNG:  breche
Verbindung ab wegen Absturz eines anderen Serverprozesses
2017-05-04 13:52:35 CEST [11095-4] postgres@hostname DETAIL:  Der
Postmaster hat diesen Serverprozess angewiesen, die aktuelle Transaktion
zurückzurollen und die Sitzung zu beenden, weil ein anderer
Serverprozess abnormal beendet wurde und möglicherweise das Shared
Memory verfälscht hat.
2017-05-04 13:52:35 CEST [11095-5] postgres@hostname TIPP:  In einem
Moment sollten Sie wieder mit der Datenbank verbinden und Ihren Befehl
wiederholen können.
```

**Comment by kou** [comment]
It seems that it's occurred when your PostgreSQL client closes session.
Could you confirm that it's always occurred or it's occurred only when some SQL is executed in the session?

**Comment by kou** [comment]
> Postgres 9.5.4 on Ubuntu 14.04 x64

16.04?

**Comment by peter-schmitz** [comment]
> Could you confirm that it's always occurred or it's occurred only when some SQL is executed in the session?
I'll look into it tomorrow. 

And it's good old Trusty 14.04 with an updated postgresql.

**Comment by peter-schmitz** [comment]
> It seems that it's occurred when your PostgreSQL client closes session.

Opened a postgresql cli, entered a query that takes a few seconds to run and uses pgroonga, then quickly aborted via CTRL-C, CTRL-D before it was done. Tried several times, but could not force an error.
I'm open to suggestions on how to test this differently/better.

Yesterday I dropped both indices for 24h and postgres did not go into recovery once.
Today I recreated both indices, then noticed the following log message an hour later:

**org.postgresql.util.PSQLException: ERROR: tuple offset out of range: 0**

This message has been logged several times since and from the second occurence onwards, postgres did always go into recovery around 1-3 Minutes after this error was thrown. It was always thrown for the only query that uses above pgroonga indices.

**Comment by peter-schmitz** [comment]
Found this in the postgres-log from yesterday now too (belongs to the first Post):

`2017-05-04 13:51:24 CEST [8378-1] postgres@hostname FEHLER:  tuple offset out of range: 0
2017-05-04 13:51:24 CEST [8378-2] postgres@hostname ANWEISUNG:  [The only query using pgroonga] 
 ...((_c.fulltext like $2) or (_c.fulltext like $3))...`

**Comment by kou** [comment]
Can you identify the SQL that causes the `tuple offset out of range: 0` error?
And could you provide the SQL and data that reproduces the problem? (Your production data aren't required. Dummy data are OK. Reproducibility is the important.)

**Comment by peter-schmitz** [comment]
>Can you identify the SQL that causes the tuple offset out of range: 0 error?

Yes. 

```text
2017-05-04 13:51:24 CEST [8378-1] postgres@hostname FEHLER:  tuple offset out of range: 0
2017-05-04 13:51:24 CEST [8378-2] postgres@hostname ANWEISUNG:  /* sq.fs=100 */select _a.produkt
          from produktsichtbarkeit _a
          join produkt _b on _a.produkt = _b.id
          join searchentry _c on _b.id = _c.produkt
         where _a.fuerendkunde and _a.webshop=$1 and _a.suchgewichtung is not null and (_c.suchklasse=10 or _c.suchklasse=20) and ((_c.fulltext1 like $2) or (_c.fulltext2 like $3))
         order by _a.suchgewichtung asc
         limit 500```

$2 and $3: "%k181956%"

One searchentry containing the word "k181956" was modified a few minutes prior - with additional words being added to fulltext1, but the searched term existed in fulltext1 before too.

Earlier Entries (than posted above) from  pgroonga.log:
```text
2017-05-04 13:50:25.213875|n|8287: grn_fin (0)
2017-05-04 13:51:17.243436|i|8283: <0x7f115ec832a0:0> expired i=0x5630ddac1c40 max=325 (41/41)
2017-05-04 13:51:17.261609|i|8283: <0x7f115ec832a0:0> expired i=0x5630dda9af40 max=617 (90/90)
2017-05-04 13:51:24.787681|i|8283: [object][search][index][key][exact] <Lexicon11521365_0.index>
2017-05-04 13:51:24.787707|i|8283: grn_ii_sel > (k181956)
2017-05-04 13:51:24.787772|i|8283: n=5 (k181956)
2017-05-04 13:51:24.788201|d|8283: [ii][cursor][min] skip: 0x5630de33ae10: min(164254->259698): chunk(2->3): chunk-used(true->true)
2017-05-04 13:51:24.788271|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(164254->259698): chunk(3->5): chunk-used(true->true)
2017-05-04 13:51:24.788680|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(404550->493035): chunk(10->11): chunk-used(true->true)
2017-05-04 13:51:24.789307|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(712874->869097): chunk(17->20): chunk-used(true->true)
2017-05-04 13:51:24.789762|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1031444->1076064): chunk(25->27): chunk-used(true->true)
2017-05-04 13:51:24.790000|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1110417->1218304): chunk(29->32): chunk-used(true->true)
2017-05-04 13:51:24.790074|d|8283: [ii][cursor][min] skip: 0x5630de33ae10: min(1218304->1435717): chunk(16->18): chunk-used(true->true)
2017-05-04 13:51:24.790097|d|8283: [ii][cursor][min] skip: 0x5630dea11810: min(1218304->1435717): chunk(13->14): chunk-used(true->true)
2017-05-04 13:51:24.790124|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1218304->1435717): chunk(33->40): chunk-used(true->true)
2017-05-04 13:51:24.790258|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1435717->1522088): chunk(41->46): chunk-used(true->true)
2017-05-04 13:51:24.790340|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1522088->1555470): chunk(47->49): chunk-used(true->true)
2017-05-04 13:51:24.790438|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1558224->1577445): chunk(50->51): chunk-used(true->true)
2017-05-04 13:51:24.790697|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1627199->1710551): chunk(54->56): chunk-used(true->true)
2017-05-04 13:51:24.791275|d|8283: [ii][cursor][min] skip: 0x5630de33ae10: min(1890677->2050035): chunk(26->27): chunk-used(true->true)
2017-05-04 13:51:24.791341|d|8283: [ii][cursor][min] skip: 0x5630dea11810: min(1890677->2050035): chunk(21->22): chunk-used(true->true)
2017-05-04 13:51:24.791369|d|8283: [ii][cursor][min] skip: 0x5630dec51470: min(1890677->2050035): chunk(63->65): chunk-used(true->true)
2017-05-04 13:51:24.791503|i|8283: exact: 14
2017-05-04 13:51:24.791510|i|8283: hits=14
2017-05-04 13:51:24.795946|i|8378: [object][search][index][key][exact] <Lexicon11521365_0.index>
2017-05-04 13:51:24.795962|i|8378: grn_ii_sel > (k181956)
2017-05-04 13:51:24.796008|i|8378: n=5 (k181956)
2017-05-04 13:51:24.796261|d|8378: [ii][cursor][min] skip: 0x5630ddba3d80: min(164254->259698): chunk(2->3): chunk-used(true->true)
2017-05-04 13:51:24.796312|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(164254->259698): chunk(3->5): chunk-used(true->true)
2017-05-04 13:51:24.796595|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(404550->493035): chunk(10->11): chunk-used(true->true)
2017-05-04 13:51:24.797011|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(712874->869097): chunk(17->20): chunk-used(true->true)
2017-05-04 13:51:24.797331|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1031444->1076064): chunk(25->27): chunk-used(true->true)
2017-05-04 13:51:24.797492|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1110417->1218304): chunk(29->32): chunk-used(true->true)
2017-05-04 13:51:24.797539|d|8378: [ii][cursor][min] skip: 0x5630ddba3d80: min(1218304->1435717): chunk(16->18): chunk-used(true->true)
2017-05-04 13:51:24.797552|d|8378: [ii][cursor][min] skip: 0x5630ddb749a0: min(1218304->1435717): chunk(13->14): chunk-used(true->true)
2017-05-04 13:51:24.797568|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1218304->1435717): chunk(33->40): chunk-used(true->true)
2017-05-04 13:51:24.797661|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1435717->1522088): chunk(41->46): chunk-used(true->true)
2017-05-04 13:51:24.797710|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1522088->1555470): chunk(47->49): chunk-used(true->true)
2017-05-04 13:51:24.797766|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1558224->1577445): chunk(50->51): chunk-used(true->true)
2017-05-04 13:51:24.797939|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1627199->1710551): chunk(54->56): chunk-used(true->true)
2017-05-04 13:51:24.798332|d|8378: [ii][cursor][min] skip: 0x5630ddba3d80: min(1890677->2050035): chunk(26->27): chunk-used(true->true)
2017-05-04 13:51:24.798370|d|8378: [ii][cursor][min] skip: 0x5630ddb749a0: min(1890677->2050035): chunk(21->22): chunk-used(true->true)
2017-05-04 13:51:24.798388|d|8378: [ii][cursor][min] skip: 0x5630ddbb4d00: min(1890677->2050035): chunk(63->65): chunk-used(true->true)
2017-05-04 13:51:24.798507|i|8378: exact: 14
2017-05-04 13:51:24.798515|i|8378: hits=14
2017-05-04 13:51:24.884149|d|8283: flushing! b=0x7f1151de7000 free=0, seg(10)
2017-05-04 13:51:24.925031|d|8283: flushed!  b=0x7f114c40b000 free=249568, seg(10)
2017-05-04 13:51:30.912366|i|8313: [object][search][index][key][exact] <Lexicon11521365_0.index>
2017-05-04 13:51:30.912383|i|8313: grn_ii_sel > (181956)
2017-05-04 13:51:30.912418|i|8313: n=4 (181956)
2017-05-04 13:51:30.914867|d|8313: [ii][cursor][min] skip: 0x5630dd59ac70: min(1222078->1303320): chunk(33->36): chunk-used(true->true)
2017-05-04 13:51:30.915257|d|8313: [ii][cursor][min] skip: 0x5630dd59ac70: min(1479703->1501695): chunk(43->44): chunk-used(true->true)
2017-05-04 13:51:30.916886|i|8313: exact: 40
2017-05-04 13:51:30.916892|i|8313: hits=40
2017-05-04 13:51:30.920866|i|8378: [object][search][index][key][exact] <Lexicon11521365_0.index>
2017-05-04 13:51:30.920884|i|8378: grn_ii_sel > (181956)
2017-05-04 13:51:30.920925|i|8378: n=4 (181956)
2017-05-04 13:51:30.923152|d|8378: [ii][cursor][min] skip: 0x5630de53dde0: min(1222078->1303320): chunk(33->36): chunk-used(true->true)
2017-05-04 13:51:30.923530|d|8378: [ii][cursor][min] skip: 0x5630de53dde0: min(1479703->1501695): chunk(43->44): chunk-used(true->true)
2017-05-04 13:51:30.925069|i|8378: exact: 40
2017-05-04 13:51:30.925079|i|8378: hits=40
2017-05-04 13:51:30.925221|i|8378: [object][search][index][key][exact] <Lexicon11521366_0.index>
2017-05-04 13:51:30.925227|i|8378: grn_ii_sel > (181956)
2017-05-04 13:51:30.925274|i|8378: n=4 (181956)
2017-05-04 13:51:30.914867|d|8313: [ii][cursor][min] skip: 0x5630dd59ac70: min(1222078->1303320): chunk(33->36): chunk-used(true->true)
2017-05-04 13:51:30.915257|d|8313: [ii][cursor][min] skip: 0x5630dd59ac70: min(1479703->1501695): chunk(43->44): chunk-used(true->true)
2017-05-04 13:51:30.916886|i|8313: exact: 40
2017-05-04 13:51:30.916892|i|8313: hits=40
2017-05-04 13:51:30.920866|i|8378: [object][search][index][key][exact] <Lexicon11521365_0.index>
2017-05-04 13:51:30.920884|i|8378: grn_ii_sel > (181956)
2017-05-04 13:51:30.920925|i|8378: n=4 (181956)
2017-05-04 13:51:30.923152|d|8378: [ii][cursor][min] skip: 0x5630de53dde0: min(1222078->1303320): chunk(33->36): chunk-used(true->true)
2017-05-04 13:51:30.923530|d|8378: [ii][cursor][min] skip: 0x5630de53dde0: min(1479703->1501695): chunk(43->44): chunk-used(true->true)
2017-05-04 13:51:30.925069|i|8378: exact: 40
2017-05-04 13:51:30.925079|i|8378: hits=40
2017-05-04 13:51:30.925221|i|8378: [object][search][index][key][exact] <Lexicon11521366_0.index>
2017-05-04 13:51:30.925227|i|8378: grn_ii_sel > (181956)
2017-05-04 13:51:30.925274|i|8378: n=4 (181956)
2017-05-04 13:51:30.931706|i|8378: exact: 53
2017-05-04 13:51:30.931726|i|8378: hits=53
2017-05-04 13:51:41.947700|n|11095: grn_init: <7.0.2>
2017-05-04 13:51:41.947824|n|11095: pgroonga: initialize: <1.2.0>
...
[crash from post #1]
```

>And could you provide the SQL and data that reproduces the problem?

I'll look into it.

**Comment by kou** [comment]
> One searchentry containing the word "k181956" was modified a few minutes prior - with additional words being added to fulltext1, but the searched term existed in fulltext1 before too.

Thanks. I've added checks for the case. It may resolve the error message.

> > And could you provide the SQL and data that reproduces the problem?
>
> I'll look into it.

Thanks. It's very helpful to confirm that the checks really resolve the error message.

**Comment by peter-schmitz** [comment]
We compiled PGroonga from this repo including your commits 2 days ago and deployed it into the same environment as the previous version.
Its now been 48 hours later and everything works flawlessly. The crashes are gone.

**Comment by kou** [comment]
Thanks for confirming the changes.

BTW, did you run `ALTER EXTENSION pgroonga UPDATE;` after `make install` as described at https://pgroonga.github.io/upgrade/#compatible-case ?

The current master has many improvements and I hope the current master keeps backward compatibility for upgrade case. But I didn't complete test for 1.2.0 -> master upgrade case yet. If `ALTER EXTENSION pgroonga UPDATE;` works well on your environment, I can release a new version with confidence.

**Comment by peter-schmitz** [comment]
> BTW, did you run ALTER EXTENSION pgroonga UPDATE; after make install ...

Sorry for not answering earlier, a colleague did the actual compiling and such. He did not run 'alter extension' and there weren't any groonga indizes active while he replaced the .so files. But we know the new files are being used due to a manually increased version number (1.2.0.x) in the log.

But I just did the same upgrade on another machine with the 1.2.0 Version and I found now problems using the following steps:
- Postgresql Service is running
- Indexes are active
- Copied our compilation "pgroonga.so" and "pgroonga_check.so" into " /usr/lib/postgresql/9.5/lib"
- Test -> Successful, the new Version number is logged to pgroonga.log.
- alter extension -> Additional test, still successful.

Postgres seems to use the new Version even without running 'alter extension'.

tl;dr did not use the above steps before, a quick local test using an even dirtier update seemed successful.
